finality-aleph/src/sync/handler/request_handler.rs

```diff
@@ -0,0 +1,423 @@
+
+            state = match branch_knowledge {
+                BranchKnowledge::LowestId(id) if *id == header.id() => State::Everything,
+                _ => state,
+            };
+
+            if let HeadOfChunk::Justification(_, _, _) = &head_chunk {
+                break;
+            }
+        }
+
+        Ok((state, head_chunk, pre_chunk.into_chunks()))
+    }
+
+    fn chunks(
+        self,
+        head: HeadOfChunk<B, J>,
+        branch_knowledge: BranchKnowledge<J>,
+        to: BlockIdFor<J>,
+    ) -> HandlerResult<Vec<Chunk<B, J>>, CS::Error> {
+        let mut chunks = vec![];
+
+        let mut head = head;
+        let mut state = State::EverythingButHeader;
+
+        while state != State::Finished {
+            let (new_state, new_head, chunks_from_step) =
+                self.step(state, head, &to, &branch_knowledge)?;
+
+            chunks.extend(chunks_from_step);
+            state = new_state;
+            head = new_head;
+        }
+
+        Ok(chunks)
+    }
+
+    pub fn response(self, request: Request<J>) -> HandlerResult<Response<B, J>, CS::Error> {
+        if !request.is_valid() {
+            return Err(RequestHandlerError::BadRequest);
+        }
+
+        let our_top_justification = self.chain_status.top_finalized()?;
+        let top_justification = request.state().top_justification();
+        let target = request.target_id();
+
+        let upper_limit = self.upper_limit(top_justification.id());
+
+        // request too far into future
+        if target.number() > upper_limit {
+            return Ok(Response::from_chunks(vec![]));
+        }
+
+        let head = match self.chain_status.status_of(target.clone())? {
+            BlockStatus::Unknown => return Ok(Response::request_block(target.clone())),
+            BlockStatus::Justified(justification) => {
+                let b = self.get_block(justification.header().id())?;
+                HeadOfChunk::Justification(justification, b, None)
+            }
+            BlockStatus::Present(header) => HeadOfChunk::Header(header),
+        };
+
+        let our_top_justification_number = our_top_justification.header().id().number();
+
+        // set the target to either target, our last justified block or finalized block
+        // at upper_limit height.
+        let head = if target.number() > our_top_justification_number {
+            head
+        } else if upper_limit > our_top_justification_number {
+            let b = self.get_block(our_top_justification.header().id())?;
+            HeadOfChunk::Justification(our_top_justification, b, None)
+        } else {
+            match self.chain_status.finalized_at(upper_limit)? {
+                FinalizationStatus::FinalizedWithJustification(j) => {
+                    let b = self.get_block(j.header().id())?;
+                    HeadOfChunk::Justification(j, b, None)
+                }
+                FinalizationStatus::FinalizedByDescendant(header) => HeadOfChunk::Header(header),
```

Comment:
This doesn't make sense, we cannot send such a chunk. On the other hand, this should never happen, as `upper_limit` is the end of a session, right? Maybe just return an error although, ugh, it will never really get returned...

Reply:
I love that it works, I hate the reasons...

Reply:
now this variant throws error which should not ever be returned